apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,222 @@
+        private bool HandleException(IEnumerable<T> items, RpcException e)
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        return false;
+                    case ExceptionHandling.Ignore:
+                        Debug.WriteLine($"Ignored exception: {e}");
+                        return true;
+                    default:
+                        Debug.Fail($"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
+                        return true;
+                }
+            }
+            
+            // Get the size of items to be buffered and retried.
+            int size = items.Sum(item => _sizer(item));
+
+            // If the new items would be greater than the size of the buffer drop them or throw.
+            if (size + _sizeInBytes > _options.BufferSizeBytes) 
+            {
+                switch (_options.BufferOverflow)
+                {
+                    case BufferOverflow.IgnoreNewEntries:
+                        Debug.WriteLine($"Ignored {items.Count()} entries.");
+                        return true;
+                    case BufferOverflow.ThrowException:
+                        throw new InvalidOperationException("Buffer overflow", e);
+                    default:
+                        Debug.Fail($"Invalid BufferOverflow option: {_options.BufferOverflow}");
+                        return true;
+                }
+            }
+
+            // Lock the semaphore and add the new items to the buffer.
+            lock (_mutex)
+            {
+                _sizeInBytes += size;
+                _buffer.AddRange(items);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Try to flush the buffer if it has any entries.
+        /// </summary>
+        private void FlushBuffer()
+        {
+            // If the buffer is empty do nothing.
+            if (_buffer.Count() == 0)
+            {
+                return;
+            }
+
+            // Move the current buffer into a new buffer to send.
+            List<T> toSend;
+            int toSendSize;
+            lock (_mutex)
+            {
+                if (_buffer.Count() == 0)
+                {
+                    return;
+                }
+                toSend = _buffer;
+                toSendSize = _sizeInBytes;
+                _buffer = new List<T>();
+                _retries++;
+            }
+
+            try
+            {
+                _consumer.Receive(toSend);
+
+                // The call succeeded, remove update the buffer size and retries.  
+                lock (_mutex)
+                {
+                    _sizeInBytes -= toSendSize;
+                    _retries = 0;
```

Comment:
The management of this `_retries` value and the general approach makes me a bit nervous. Do we need to worry about the order of RPCs here? What happens if thread A increments the count above, clears the buffer, and is about to send the content to the consumer, but then on thread B, another item is added to the buffer and it gets flushed before thread A can actually finish? Should this whole method be in a single lock statement?